spring-integration-core/src/main/java/org/springframework/integration/transformer/ToProtobufTransformer.java

```diff
@@ -0,0 +1,104 @@
+import org.springframework.util.Assert;
+
+/**
+ * An Protocol Buffer transformer for generated {@link com.google.protobuf.Message} objects.
+ * 
+ * @author Christian Tzolov
+ * @since 6.1
+ */
+public class ToProtobufTransformer extends AbstractTransformer {
+
+    private Expression typeIdExpression = new FunctionExpression<Message<?>>(
+            (message) -> message.getPayload().getClass());
+
+    private EvaluationContext evaluationContext;
+
+    /**
+     * Set the expression to evaluate against the message to determine the value for the {@link ProtoHeaders#TYPE}
+     * header.
+     * @param expression the expression.
+     * @return the transformer
+     */
+    public ToProtobufTransformer typeExpression(Expression expression) {
+        assertExpressionNotNull(expression);
+        this.typeIdExpression = expression;
+        return this;
+    }
+
+    /**
+     * Set the expression to evaluate against the message to determine the value for the {@link ProtoHeaders#TYPE}
+     * header.
+     * @param expression the expression.
+     * @return the transformer
+     */
+    public ToProtobufTransformer typeExpression(String expression) {
+        assertExpressionNotNull(expression);
+        this.typeIdExpression = EXPRESSION_PARSER.parseExpression(expression);
+        return this;
+    }
+
+    /**
+     * Set the expression to evaluate against the message to determine the value for the {@link ProtoHeaders#TYPE}
+     * header.
+     * @param expression the expression.
+     */
+    public void setTypeExpression(Expression expression) {
+        assertExpressionNotNull(expression);
+        this.typeIdExpression = expression;
+    }
+
+    /**
+     * Set the expression to evaluate against the message to determine the value for the {@link ProtoHeaders#TYPE}
+     * header.
+     * @param expression the expression.
+     */
+    public void setTypeExpressionString(String expression) {
+        assertExpressionNotNull(expression);
+        this.typeIdExpression = EXPRESSION_PARSER.parseExpression(expression);
+    }
+
+    private void assertExpressionNotNull(Object expression) {
+        Assert.notNull(expression, "'expression' must not be null");
+    }
+
+    @Override
+    protected void onInit() {
+        this.evaluationContext = IntegrationContextUtils.getEvaluationContext(getBeanFactory());
+    }
+
+    @Override
+    protected Object doTransform(Message<?> message) {
+        Assert.state(message.getPayload() instanceof com.google.protobuf.Message,
+                "Payload must be an implementation of 'com.google.protobuf.Message'");
+        com.google.protobuf.Message protobufMessage = (com.google.protobuf.Message) message.getPayload();
+
+        return getMessageBuilderFactory().withPayload(protobufMessage.toByteArray())
+                .copyHeaders(message.getHeaders())
+                .setHeader(ProtoHeaders.TYPE, this.typeIdExpression.getValue(this.evaluationContext, message))
```

Comment:
Do you mean we cannot infer the proto type from that `com.google.protobuf.Message`?
How can we be sure that expression will return a proper type we have just deserialzed from?

Reply:
I think you still have a question here.
So, what I see so far: since `FromProtobufTransformer` relies on a `com.google.protobuf.Message.class` in that `ProtoHeaders.TYPE`, then I don't see a reason in such an extra expression and just provide whatever info is our `payload` based on.
It is probably up to target `FromProtobufTransformer` (or any other consumer) side to decide to rely on our `ProtoHeaders.TYPE` header or use some other  type for Proto deserialization.

WDYT if we just remove this `typeIdExpression` altogether? 

I see that you took this logic from the `SimpleToAvroTransformer`, but I still wonder why would one use different type for header value from the one we have just serialized from?...